bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,401 @@
+use std::{
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+use s2n_tls_sys::*;
+
+use crate::error::{Error, Fallible};
+
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+// this is the size of the MD5 hash digest that is used for the JA3 fingerprint
+const MD5_HASH_SIZE: u32 = 16;
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+/// ClientHello is an opaque wrapper struct around `s2n_client_hello`. Note that
+/// the size of this type is not known, and as such it can only be used through
+/// references and pointers.
+///
+/// This implementation is motivated by the different memory_management required
+/// for different s2n_client_hello pointers. `s2n_client_hello_parse_message`
+/// returns a `*mut s2n_client_hello` which owns it's own data. This neatly fits
+///  the "smart pointer" pattern and can be represented as a `Box<T>`.
+///
+/// `s2n_connection_get_client_hello` returns a `*mut s2n_client_hello` which
+/// references memory owned by the connection, and therefore must not outlive
+/// the connection struct. This is best represented as a reference tied to the
+/// lifetime of the `Connection` struct.
+
+pub struct ClientHello(s2n_client_hello);
+
+// safety justifications
+// 1 - casting *s2n_client_hello <-> *ClientHello: a struct with only one field
+// has the same layout as that field
+// https://rust-lang.github.io/unsafe-code-guidelines/layout/structs-and-tuples.html#single-field-structs
+// 2 - casting *const s2n_client_hello -> *mut s2n_client_hello: This is safe as
+// long as the data is not actually mutated. As authors of s2n-tls, we know that
+// the get_hash and get_fingerprint methods do not mutate the data, and use mut
+// pointers as a matter of convention because it makes working with s2n_stuffers
+// and s2n_blobs easier.
+
+impl Deref for ClientHello {
+    type Target = s2n_client_hello;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for ClientHello {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl ClientHello {
+    pub fn parse_client_hello(hello: &[u8]) -> Result<Box<Self>, crate::error::Error> {
+        let handle = unsafe {
+            s2n_client_hello_parse_message(hello.as_ptr(), hello.len() as u32).into_result()?
+        };
+        let client_hello = handle.as_ptr() as *mut ClientHello;
+        // safety: s2n_client_hello_parse_message returns a pointer that "owns"
+        // it's memory. This memory must be cleaned up by the application. The
+        // Box<Self> will call Self::Drop when it goes out of scope so memory
+        // will be automatically managed.
+        unsafe { Ok(Box::from_raw(client_hello)) }
```

Comment:
I don't think this is sound... We're allocating the client hello through s2n-tls but freeing it through `Box`, meaning the layouts/alignment could potentially be different.

Reply:
I don't think the layout is a concern, since `s2n_client_hello` is an opaque struct. My understanding is also that bindgen automatically matches alignments for opaque types.
> Opaque in this context means that none of the generated bindings will contain information about the inner representation of the type and the type itself will be represented as a chunk of bytes with the alignment and size of the type.

https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#method.opaque_type